### ISSUE TITLE
Sniff for HTML in translatable strings

### ISSUE THREAD

**Issue by jrfnl** [WordPress/WordPress-Coding-Standards#1419]
> Avoid HTML in strings if possible

Ref: https://make.wordpress.org/core/handbook/best-practices/internationalization/

HTML in translatable strings should be avoided as it puts an extra burden on translators in that they'd have to know HTML.

It also makes output escaping translatable strings much easier if there is no HTML in the string.

The HTML should be moved out of the translatable string and placeholders should be used instead.
```php
// Bad:
echo wp_kses( __( 'This is a <strong>bold</strong> text', 'text-domain' ) );

// Good:
printf (
    /* translators: 1: <strong> open tag 2: close tag. */
    esc_html__( 'This is a %1$sbold%2$s text', 'text-domain' ),
    '<strong>',
    '</strong>'
);
```

I think this should be quite easily sniffable and could be a good addition to the existing `WP.I18n` sniff.

I propose for this to be a `warning`.

I also propose *not* to include an auto-fixer. While for a simple example as the above, this could be auto-fixed, in general, it will need the critical look of a human being to determine the best course of action.

**Comment by aristath** [comment]
Just a question out of curiosity

```php
// Good:
printf (
    /* translators: 1: <strong> open tag 2: close tag. */
    esc_html__( 'This is a %1$sbold%2$s text', 'text-domain' ),
    '<strong>',
    '</strong>'
);
```

Would this not be better like this since it's easier for translators?
```
// Good:
printf (
    /* translators: The "bold" string, wrapped in <strong> tags. */
    esc_html__( 'This is a %s text', 'text-domain' ),
    '<strong>' . esc_html__( 'bold', 'text-domain' ) . '</strong>',
);
```

Strings like `%1$sbold%2$s` are something that I've seen confuse lots of people and over the years I've spotted many wrong translations because of strings like that

**Comment by jrfnl** [comment]
@aristath No. You should always avoid fragmenting a sentence as the sentence build up may be (very) different in different languages.

See: https://make.wordpress.org/core/handbook/best-practices/internationalization/#avoid-fragmented-or-patchwork-messages

**Comment by aristath** [comment]
Ah I don't know how I missed that... Thanks! I learned something new today. 👍

**Comment by GaryJones** [comment]
I'm -1 for this.

The [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers#Best_Practices) say to move _unusual_ markup out of the string, but it also says about leaving URLs (typically, as part of an `<a>` tag) in if they are likely to have a locale-specific version.

It then also refers to the [`gettext` manual](https://www.gnu.org/software/gettext/manual/html_node/Preparing-Strings.html#Preparing-Strings), which includes:

> HTML markup, however, is common enough that it’s probably ok to use in translatable strings.

So a string with basic markup is fine. It's static markup - translators can cope with it, as they have done for the last X years.

The Handbook page to which you reference is clearly incomplete, so I wouldn't put too much weight behind those 6 words to feel it implies that _all_ HTML should be removed.

The [Theme Handbook](https://developer.wordpress.org/themes/functionality/internationalization/#best-practices-for-writing-strings) clarifies it as:

> Do not put _unnecessary_ HTML markup into the translated string.
(my emphasis)

rather than do not put _any_ markup in the string.

The [Plugin Handbook](https://developer.wordpress.org/plugins/internationalization/how-to-internationalize-your-plugin/#descriptions) even gives an example with HTML in it:

```php
/* translators: 1: WordPress version number, 2: plural number of bugs. */
_n_noop( '<strong>Version %1$s</strong> addressed %2$s bug.',
         '<strong>Version %1$s</strong> addressed %2$s bugs.' );
```

Examples where the markup _should_ be removed from inside of the string would be:

```
<?php __( '<h1>Settings Page</h1>', 'text-domain ); ?>
should be
<h1><?php __( 'Settings Page', 'text-domain' ); ?></h1>
because the markup only wraps the string.
```

```
<?php _e( 'Adds <code>rel="noopener noreferer"</code> to links', 'text-domain' ); ?>
would move the code and the `<code>` elements into a placeholder, because they should not change during translation:
<?php
printf(
    esc_html__( 'Adds %s to links', 'text-domain' ),
    '<code>rel="noopener noreferer"</code>'
); ?>
```

The first example _might_ be detectable - matching tags as first and last items in the string, with no other markup, but it would be trivial to come up with close-but-not-quite right examples where a check like this would miss.

**Comment by jrfnl** [comment]
@GaryJones Good points. I think the second example would be quite easy to sniff for as well as `<code>` is a very specific tag.

Let's see if we can collect some more typical examples where HTML should be moved out of the translatable string.
We could then just throw a `warning` for those specific situations.

**Comment by ocean90** [comment]
I fully agree with @GaryJones's statements above; `%1$sbold%2$s` is definitely not _good_. 

> Let's see if we can collect some more typical examples where HTML should be moved out of the translatable string.

See https://core.trac.wordpress.org/search?q=%22Avoid+using+HTML%22&noquickjump=1&changeset=on&ticket=on for some examples.

**Comment by jrfnl** [comment]
Thanks @ocean90 for that link, very useful!

**Comment by jrfnl** [comment]
There doesn't seem to be consensus on this, so closing this issue as `wont-fix`.
